### ISSUE TITLE
Buffer underflow with several parallel requests

### ISSUE THREAD

**Issue by simendsjo** [SuaveIO/suave#379]
Suave 1.0 from NuGet on .NET 4.5.0 and F# 4.0

I found an issue when adding a parallel testcase in my project. This is only triggered when I spawn many parallel requests (30+). I don't have time to debug at the moment, but I'm adding the issue here in case it's a known issue. `Suave.ParsingAndControl.readMoreData` is called when the buffer is empty (is my first guess).

```
[W] 2016-01-18T16:01:04.6927634Z: tcp request processing failed [Suave.Tcp.job] exn:
System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.Stack`1.Pop()
   at Suave.Sockets.BufferManager.PopBuffer(FSharpOption`1 context) in Z:\ademar\Projects\suave\src\Suave\Sockets\BufferManager.fs:line 26
   at Suave.ParsingAndControl.readMoreData@113.Invoke(Unit unitVar) in Z:\ademar\Projects\suave\src\Suave\ParsingAndControl.fs:line 113
   at Microsoft.FSharp.Control.AsyncBuilderImpl.callA@851.Invoke(AsyncParams`1 args)
```

EDIT: I'll try to create a smaller testcase in the following days if necessary.

**Comment by haf** [comment]
Might be a duplicate of #363

**Comment by simendsjo** [comment]
Increasing `maxOps` as noted in #363 makes the buffer underflow issue disappear. Seems like a hack though.

**Comment by haf** [comment]
It's not a hack, but how computers work with finite resources. ;) 

The fix [is easy](https://github.com/SuaveIO/suave/issues/363#issuecomment-172457140) – if you're going to use this in production, consider spending an hour to solve it? We'd be grateful for your contribution.

**Comment by simendsjo** [comment]
I was thinking about me just increasing `maxOps` as the hack. I don't know the number of concurrent clients, so I'll just end up pushing this too far up. I'll see if I can take a stab at it during the weekend.

**Comment by ronnieholm** [comment]
@simendsjo: You can override the default Suave config to make Suave log the number of concurrent requests and how much buffer space is left:

    startWebServer { defaultConfig with maxOps = 10; logger = Logging.Loggers.saneDefaultsFor Logging.LogLevel.Verbose } webPart

The buffer space isn't that large: 827392 bytes for the default maxOps of 100.

**Comment by haf** [comment]
Fix released as v1.1
